File: services/api/metrics.go
package main

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	metricAPIRequestsTotal = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "api_requests_total",
		Help: "HTTP requests handled, by method, path, and status.",
	}, []string{"method", "path", "status"})

	metricAPIRequestDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
		Name:    "api_request_duration_seconds",
		Help:    "HTTP request handling duration, by method and path.",
		Buckets: prometheus.DefBuckets,
	}, []string{"method", "path"})

	metricWebsocketConnections = promauto.NewGauge(prometheus.GaugeOpts{
		Name: "websocket_connections",
		Help: "Currently connected WebSocket clients.",
	})
)

File: services/api/metrics.py
"""Prometheus metrics, the Python port of metrics.go. Same metric names and
labels, so existing Grafana dashboards/alerts keep working unchanged."""

from prometheus_client import Counter, Gauge, Histogram

api_requests_total = Counter(
    "api_requests_total",
    "HTTP requests handled, by method, path, and status.",
    ["method", "path", "status"],
)

api_request_duration_seconds = Histogram(
    "api_request_duration_seconds",
    "HTTP request handling duration, by method and path.",
    ["method", "path"],
)

websocket_connections = Gauge(
    "websocket_connections",
    "Currently connected WebSocket clients.",
)

devices_by_status = Gauge(
    "devices_by_status",
    "Devices grouped by lifecycle status, across all organizations -- backs the IoT dashboard's active/offline device panels.",
    ["status"],
)
